jaxrs-api/src/main/java/javax/ws/rs/ext/MessageBodyWriter.java

```diff
@@ -27,7 +27,8 @@
  * Contract for a provider that supports the conversion of a Java type to a stream.
  *
  * A {@code MessageBodyWriter} implementation may be annotated with {@link javax.ws.rs.Produces} to restrict the media
- * types for which it will be considered suitable.
+ * types for which it will be considered suitable. The {@code MessageBodyWriter} pipeline is only invoked if there is 
```

Comment:
Lessions learned from other PRs: The term "there is" might lead to discussions about the meaning of `null` in future, as one could say "there was an entity, but it was null" vs "The invoked method has a return type of `void`". So maybe it would be more unambiguous if we say "...is only invoked if the invoked resource method returned a non-`null` value"?

Reply:
I agree that using the term "non-null" is much better and I updated the pull request accordingly. See the updated version.

But I don't think that we should refer to the return value of the resource method directly. Mostely, because there are other components (like `ContainerResponseFilter`) which can alter the response entity as part of the processing chain. In the end it just matters what the response entity is just before the MBW chain is invoked. And at this point the entity could differ from the return value of the resource (if somebody called `ContainerResponseContext.setEntity()` for example).

Reply:
I do not see what components you think have the power to prevent the invocation *of the MBW* according to the JAX-RS 2.1 spec text and JavaDocs:
* A `ContainerResponseFilter` is invoked *after* an MBW is invoked, so this is irrelevant for your PR, as even a `null`-result of the filter could not prevent the invocation *of the MBW*.
* A `WriterInterceptor` is invoke itself *only if an MBW* is invoked as it is "wraps around calls to MBW#write"* according to its JavaDocs, so even if the interceptor produces `null` this would not stop *the MBW* from being called.

Reply:
No, `ContainerResponseFilter` is invoked after the resource, but before the MBW. And `ContainerResponseFilter.setEntity(Object)` effectively allows to "change" the entity returned from the resource.

![img](https://d1ldz4te4covpm.cloudfront.net/graphics/9781788294041/graphics/f0d3b0d5-3dba-4cb3-a216-4476df4cf3b1.png)

Reply:
LOL, I'm participating to JAX-RS spec since more than a decade but actually learn new things every day! ;-)

While it is totally weird to have response filters being called *before* the writer, I now understand your argument!

Reply:
@chkal @mkarg There are some diagrams about this in Appendix C in the spec document. I think those are a bit more complete. It is indeed the case that response filters are called before the MBW chain.